msbuild/Xamarin.MacDev.Tasks/Tasks/AnalyzeFileChanges.cs

```diff
@@ -0,0 +1,97 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Xamarin.Messaging;
+using Xamarin.Messaging.Build.Client;
+
+#nullable disable
+
+namespace Xamarin.MacDev.Tasks {
+	public class AnalyzeFileChanges : XamarinTask, ITaskCallback {
+		[Required]
+		public string WorkingDirectory { get; set; } = string.Empty;
+
+		[Required]
+		public ITaskItem HashesFile { get; set; }
+
+		[Output]
+		public ITaskItem [] ChangedFiles { get; set; } = [];
+
+		public IEnumerable<ITaskItem> GetAdditionalItemsToBeCopied () => [];
+
+		//We need the HashesFile input to be copied to the Mac. Since it's the only ITaskItem input, it's safe to return true
+		public bool ShouldCopyToBuildServer (ITaskItem item) => true;
+
+		//In case it's a remote execution, we don't want empty output files to be copied since we need the real files to be copied
+		public bool ShouldCreateOutputFile (ITaskItem item) => false;
+
+		public override bool Execute ()
+		{
+			if (ShouldExecuteRemotely ()) {
+				return new TaskRunner (SessionId, BuildEngine4).RunAsync (this).Result;
+			}
+
+			if (!Directory.Exists (WorkingDirectory)) {
+				Log.LogError ("Unable to analyze file changes in '{0}'. Directory not found.", WorkingDirectory);
+
+				return false;
+			}
+
+			if (!File.Exists (HashesFile.ItemSpec)) {
+				Log.LogError ("Unable to analyze file changes in '{0}'. The hashes file '{1}' does not exist", WorkingDirectory, HashesFile.ItemSpec);
+
+				return false;
+			}
+
+			try {
+				var changedFiles = new List<ITaskItem> ();
+				//Method from Xamarin.Messaging.Common
+				using HashAlgorithm hashAlgorithm = Hash.GetAlgorithm ();
+				//Gets a dictionary of file names and hashes from the HashesFile
+				IDictionary<string, string> fileHashesList = GetFileHashesList ();
+
+				foreach (string file in Directory.GetFiles (WorkingDirectory, "*", SearchOption.TopDirectoryOnly)) {
+					//We only want the files that are present in the HashesFile
```

Comment:
Maybe also add the file size to the list of files+hashes? It's much faster to check the file size, and if the file size check doesn't match, we won't have to compute the hash.

Reply:
Done